theano/tensor/nnet/neighbours.py

```diff
@@ -111,6 +111,43 @@ def grad(self, inp, grads):
                 return [neibs2images(gz, neib_shape, x.shape, mode=self.mode),
                         grad_undefined(self, 1, neib_shape),
                         grad_undefined(self, 2, neib_step)]
+
+        if self.mode in ['valid']:
```

Comment:
The current implementation only works when the step size is equal to the
neighborhood size.

This gradient is useful for implementing neural network layers with 'local
connectivity' --- layers shaped like a convolutional layer but without the
weight-sharing. This scan implementation is reasonably fast.


On Fri, Oct 23, 2015 at 4:07 PM Frédéric Bastien <notifications@github.com>
wrote:

> In theano/tensor/nnet/neighbours.py
> <https://github.com/Theano/Theano/pull/3547#discussion_r42923582>:
>
> > @@ -111,6 +111,43 @@ def grad(self, inp, grads):
> >                  return [neibs2images(gz, neib_shape, x.shape, mode=self.mode),
> >                          grad_undefined(self, 1, neib_shape),
> >                          grad_undefined(self, 2, neib_step)]
> > +
> > +        if self.mode in ['valid']:
>
> This case is already implemented. Why do you want to reimplement it? Using
> scan for this would be very slow.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/3547/files#r42923582>.
>